### ISSUE TITLE
Weird bug with properties used both in `with` and `if`

### ISSUE THREAD

**Issue by jollygreenlaser** [edgedb/edgedb#7161]
- EdgeDB Version: EdgeDB Version: 4.4+596d0e4 on Ubuntu 22.04.3 LTS
- EdgeDB CLI Version: EdgeDB CLI 4.1.0+a8fe4d7

Here's a fun subtle bug for ya:

```
type A {
   name: str;
}

type B {
  multi a: A;
}

type C {
  required status: Status;
}

select C {
  bs := (
    with complete := .status = Status.Complete
    select B {
      complete := complete,
      a: {
        name,
        complete := complete,
      }
    }
  ) if .status != Status.Private else {}
} filter .id = <uuid>$0
```

`complete` will be `{}` in `bs` and `a` will be an empty set, no matter what value `status` is or how many of `A` are associated to that `B`.

This is the case for any property used in the `if`, as far as I can tell. 

This feels related to the normal empty set footguns, except `status` is clearly `required` and even then should still see `complete` in `a` be an empty set, rather than removing all of `a`.

**Comment by msullivan** [comment]
Could you provide a definition of Status or tweak it to not use a scalar that's missing? Thanks!

**Comment by jollygreenlaser** [comment]
@msullivan done

**Comment by msullivan** [comment]
Thanks! It looks like this is some sort of insane path factoring related bug. I'm digging into it more, but if you put the reference to `.status` in the if condition inside a subquery (`(select .status)` or `{.status}`) it prevents the issue.


A reduced version is
```
select C {
  bs := (
    with complete := .status
    select {
      complete := complete,
    }
  ) if .status != 'Private' else {}
};
```

**Comment by jollygreenlaser** [comment]
Smart. Forgot to mention, I got around it by throwing the `if` check as a part of the filter on the select. Seems to be mildly faster too.

**Comment by msullivan** [comment]
> Smart. Forgot to mention, I got around it by throwing the `if` check as a part of the filter on the select. Seems to be mildly faster too.

Yeah, that is a generally cleaner approach and will typically be faster. `FILTER` (`WHERE` in SQL) is a pretty fundamental concept in SQL, and `IF` is not.

**Comment by msullivan** [comment]
Another note: the ability to have an object with a shape in the `then` clause of an `if` and and empty set in the `else` clause and have it preserve the output shape was introduced completely by accident in 4.0. I didn't realize it worked until the issue. Normally an `if` always loses the output shape. 

Arguably it's a bug that it doesn't here.

**Comment by jollygreenlaser** [comment]
Sounds like the RFC to remove path factoring will fix this
